Added GA API version for Scheduled Query Rule

abversqr opened the pull request:
Added GA API version for Scheduled Query Rule

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

abversqr committed:
Added GA API version for Scheduled Query Rule

Added GA API version for Scheduled Query Rule

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-libraries-for-java

AutorestCI commented:
# Automation for azure-sdk-for-go

fearthecowboy commented:
Yoink! @sarangan12 -- I'm assigning this to me so I can use it to train more reviewers.  You get a freebie!

fearthecowboy commented:
Hey @abversqr  -- it appears you have a few validation errors in your spec -- can you correct these and update the PR? 

https://travis-ci.org/Azure/azure-rest-api-specs/jobs/365583431#L736

If you want to run the linter yourself locally, [install autorest](https://aka.ms/autorest/install) and try:

``` batch
# from the folder where the readme.md file is
autorest --azure-validator readme.md 
```

It looks mostly like one typo (`date=time`?) missing examples and an API that should be defined.

fearthecowboy commented:
Hey @abversqr  -- this looks like a new API is being added -- has this been reviewed by the API review board? 

If not, you should schedule a one hour SKYPE meeting with Garrett Serack, David Justice, Johan Stenberg, Gaurav Bhatnagar, Darrel Miller and Niklas Gustafsson . Preferably Tuesday afternoon sometime between 1 and 5pm.

fearthecowboy commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+  "info": {
+    "title": "Microsoft Insights API",
+    "version": "2018-04-16",
+    "x-ms-code-generation-settings": {
```
You shouldn't have `x-ms-code-generation-settings` in the actual swagger file (I don't think we even use this anymore at all).

abversqr replied:
Fixed.

fearthecowboy commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/microsoft.insights/scheduledQueryRules/{ruleName}": {
+      "put": {
+        "description": "Creates or updates an log search rule.\r\nRequest method: PUT\t\tRequest URI: https://management.azure.com/subscriptions/{subscription-id}/resourceGroups/{resource-group-name}/providers/microsoft.insights/scheduledQueryRules/{logsearch-rule-name}?api-version={api-version}",
```
The `description` should  document the API (the presence of the swagger file is telling you how you should physically call this method). You can drop everything after `Creates or updates an log search rule.`

abversqr replied:
Fixed in latest iteration

fearthecowboy commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+            "schema": {
+              "$ref": "#/definitions/LogSearchRuleResource"
+            },
+            "examples": {
```
This should use `x-ms-examples` and the example data should be in a separate file. (and then the linter won't yell at you for missing examples :D )

abversqr replied:
Fixed in latest iteration

fearthecowboy commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+        "tags": [
+          "scheduledQueryRules"
+        ],
+        "operationId": "getScheduledQueryRule",
```
again, fix the `operationId` to match the pattern (ie`ScheduledQueryRules_Get`)

fearthecowboy commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+            "schema": {
+              "$ref": "#/definitions/LogSearchRuleResource"
+            },
+            "examples": {
```
again, move example into external file with `x-ms-examples`

fearthecowboy commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+            "description": "The parameters of the rule to create or update."
+          }
+        ],
+        "responses": {
```
You should support a `"default"` response for errors. (see example https://github.com/abversqr/azure-rest-api-specs/blob/2e4c790d16eb6be9299db7a75798bca1697fd5e9/specification/monitor/resource-manager/microsoft.insights/stable/2018-03-01/metricAlert_API.json#L503 )

fearthecowboy commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+              }
+            }
+          },
+          "404": {
```
Error states should be modeled as `"default"` responses and have a model to deserialize the error into.

A lot of these things should have been caught in the API review board -- I'll pause the review here, and wait until after the review board has seen these APIs.

fearthecowboy reviewed (changes requested):
You will need to resolve a bunch of this before I can continue the review -- and the API review board should sign off.

ravbhatnagar commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+          "readOnly": true,
+          "description": "Azure resource Id"
+        },
+        "name": {
```
this should also be readonly since the name comes from URL

abversqr replied:
Fixed in latest iteration

ravbhatnagar commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+            "modelAsString": true
+          }
+        },
+        "skuType": {
```
SKUs of a resource is indicated through the ARM top level "sku" property. The details can be found in the ARM resource provider contract

ravbhatnagar commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+          "format": "date-time",
+          "description": "Last time the rule was updated in IS08601 format."
+        },
+        "provisioningState": {
```
please define as an enum and list all the possible values.

abversqr replied:
Fixed in latest iteration

abversqr committed:
Resolving Comments

Resolving Comments

abversqr commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,794 @@
+        "tags": [
+          "scheduledQueryRules"
+        ],
+        "operationId": "createOrUpdateScheduledQueryRules",
```
Fixed in latest iteration

abversqr committed:
Made SKU as top level proprty and including generic error response format

Couple of changes -
1. Made SKU as top level proprty as mentioned here - https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#put-resource
 2. including generic error response format

ravbhatnagar commented:
Looks good!

abversqr commented:
Hey @fearthecowboy, I have resolved all comments. Can ou please have a look into this and sign-off if no blocking changes are pending,

abversqr committed:
Minor Changes to align implementation

1. Removed SKU - Billing Model is still under discussion, hence need not to be exposed to customers.
2.  Enabled field ichanged to align with camel case.
3.  'throttleTillDate' in properties.action changed to 'throttlingInMin' to denote correct meaning  and it’s data type changed to number instead of date.
4. 'status' to be removed in properties.action from examples.
5.  'severity' made a required field in properties.action.
6. Removed Examples from original spec json

abversqr committed:
Updated Example to reflect AI instances in tags and 'dataSourceId"

Updated Example to reflect AI instances in tags and 'dataSourceId"

abversqr committed:
made 'tags' details more descriptive

abversqr committed:
Restore Tags Description

Restore Tags Description

abversqr committed:
Fixed Exmaple Validation

Fixed errores reported by running command "oav validate-example"

fearthecowboy commented on specification/monitor/resource-manager/microsoft.insights/stable/2018-04-16/scheduledQueryRule_API.json:
```diff
@@ -0,0 +1,628 @@
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
```
You should have a `default` response modeled here -- if the service fails for any reason, the client needs to be able to get the error response.

abversqr committed:
Added Default Response payload in all APIs

Changes -
1. Added Default Response payload in all APIs
2. Made azNs description more clear

fearthecowboy reviewed (approved):
LGTM.

There is still a R3026 Linter violation -- the RP will fix in a subsequent PR : __"we are working on this, but will take couple of weeks to get this done. Once this done, we will push this change in new PR for review. Let us know if you see any concern with this."__
